### PR TITLE
states: check root output value changes in refresh-only mode

### LINKED ISSUE #35806: Terragrunt apply -refresh-only results in Error: Cannot apply non-applyable plan
### Terraform Version

```shell
Terraform v1.9.4
on linux_amd64
```

### Terraform Configuration Files

```terraform
variable "test" {
  type = string
}

output "test" {
  value = var.test
}
```

### Debug Output

https://gist.github.com/avanschie/080d0cb2018d580ac51f2152547cf897

### Expected Behavior

I expect the refresh to succeed, updating the output in the state with the new value. When I run the deprecated `terraform refresh -var test=bar` instead of `terraform apply -refresh-only -var test=bar`, the refresh succeeds.

### Actual Behavior

In the case where the only change is the value of an output, the refresh (i.e., `terraform apply -refresh-only`) doesn't succeed and reports `Error: Cannot apply non-applyable plan`.

### Steps to Reproduce

1. `terraform init`
2. `terraform apply -auto-approve -var test=foo`
3. `terraform apply -refresh-only -var test=bar`

### Additional Context

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by bschaatsbergen**
Fixes #35806

In refresh-only mode, we don’t expect to propose any actions, but a plan is marked as “applyable” if there are state changes between runs. Currently, a plan is considered “applyable” only when there are differences in managed resources, which overlooks changes in root output values. As a result, a plan might be incorrectly marked as non-applyable even when the root output values have changed, since only managed resources are being checked.

The [states](https://github.com/hashicorp/terraform/tree/main/internal/states) package has an [Equal](https://github.com/hashicorp/terraform/blob/main/internal/states/state_equal.go#L18) function that checks if two states are deeply equal. While this could solve the issue, it seemed like an overreach, as it focuses on managed resource instances only. To address this, I added a function specifically to detect changes in root output values between states.

It’s also worth noting that output values in other modules don’t persist between runs. Terraform Core tracks these internally, but they aren’t included in any artifacts that persist across executions.

Question for the reviewer: What about ephemeral output values? And, are data sources considered as part of the existing `ManagedResourcesEqual` function already?

```
 $ go test -run "TestState" -v
=== RUN   TestState
--- PASS: TestState (0.00s)
=== RUN   TestStateDeepCopyObject
--- PASS: TestStateDeepCopyObject (0.00s)
=== RUN   TestStateDeepCopy
--- PASS: TestStateDeepCopy (0.00s)
=== RUN   TestStateHasResourceInstanceObjects
=== RUN   TestStateHasResourceInstanceObjects/empty
=== RUN   TestStateHasResourceInstanceObjects/one_current,_ready_object_in_root_module
=== RUN   TestStateHasResourceInstanceObjects/one_current,_ready_object_in_child_module
=== RUN   TestStateHasResourceInstanceObjects/one_current,_tainted_object_in_root_module
=== RUN   TestStateHasResourceInstanceObjects/one_deposed,_ready_object_in_root_module
=== RUN   TestStateHasResourceInstanceObjects/one_empty_resource_husk_in_root_module
=== RUN   TestStateHasResourceInstanceObjects/one_current_data_resource_object_in_root_module
--- PASS: TestStateHasResourceInstanceObjects (0.00s)
    --- PASS: TestStateHasResourceInstanceObjects/empty (0.00s)
    --- PASS: TestStateHasResourceInstanceObjects/one_current,_ready_object_in_root_module (0.00s)
    --- PASS: TestStateHasResourceInstanceObjects/one_current,_ready_object_in_child_module (0.00s)
    --- PASS: TestStateHasResourceInstanceObjects/one_current,_tainted_object_in_root_module (0.00s)
    --- PASS: TestStateHasResourceInstanceObjects/one_deposed,_ready_object_in_root_module (0.00s)
    --- PASS: TestStateHasResourceInstanceObjects/one_empty_resource_husk_in_root_module (0.00s)
    --- PASS: TestStateHasResourceInstanceObjects/one_current_data_resource_object_in_root_module (0.00s)
=== RUN   TestStateHasRootOutputValues
=== RUN   TestStateHasRootOutputValues/empty
=== RUN   TestStateHasRootOutputValues/one_output_value
=== RUN   TestStateHasRootOutputValues/one_secret_output_value
=== RUN   TestStateHasRootOutputValues/one_output_value_in_child_module
=== RUN   TestStateHasRootOutputValues/one_output_value_in_multi_module
--- PASS: TestStateHasRootOutputValues (0.00s)
    --- PASS: TestStateHasRootOutputValues/empty (0.00s)
    --- PASS: TestStateHasRootOutputValues/one_output_value (0.00s)
    --- PASS: TestStateHasRootOutputValues/one_secret_output_value (0.00s)
    --- PASS: TestStateHasRootOutputValues/one_output_value_in_child_module (0.00s)
    --- PASS: TestStateHasRootOutputValues/one_output_value_in_multi_module (0.00s)
=== RUN   TestState_MoveAbsResource
=== RUN   TestState_MoveAbsResource/basic_move
=== RUN   TestState_MoveAbsResource/move_to_new_module
=== RUN   TestState_MoveAbsResource/from_a_child_module_to_root
=== RUN   TestState_MoveAbsResource/module_to_new_module
=== RUN   TestState_MoveAbsResource/module_to_new_module#01
--- PASS: TestState_MoveAbsResource (0.00s)
    --- PASS: TestState_MoveAbsResource/basic_move (0.00s)
    --- PASS: TestState_MoveAbsResource/move_to_new_module (0.00s)
    --- PASS: TestState_MoveAbsResource/from_a_child_module_to_root (0.00s)
    --- PASS: TestState_MoveAbsResource/module_to_new_module (0.00s)
    --- PASS: TestState_MoveAbsResource/module_to_new_module#01 (0.00s)
=== RUN   TestState_MaybeMoveAbsResource
=== RUN   TestState_MaybeMoveAbsResource/first_move
=== RUN   TestState_MaybeMoveAbsResource/noop
--- PASS: TestState_MaybeMoveAbsResource (0.00s)
    --- PASS: TestState_MaybeMoveAbsResource/first_move (0.00s)
    --- PASS: TestState_MaybeMoveAbsResource/noop (0.00s)
=== RUN   TestState_MoveAbsResourceInstance
=== RUN   TestState_MoveAbsResourceInstance/resource_to_resource_instance
=== RUN   TestState_MoveAbsResourceInstance/move_to_new_module
--- PASS: TestState_MoveAbsResourceInstance (0.00s)
    --- PASS: TestState_MoveAbsResourceInstance/resource_to_resource_instance (0.00s)
    --- PASS: TestState_MoveAbsResourceInstance/move_to_new_module (0.00s)
=== RUN   TestState_MaybeMoveAbsResourceInstance
=== RUN   TestState_MaybeMoveAbsResourceInstance/first_move
=== RUN   TestState_MaybeMoveAbsResourceInstance/noop
--- PASS: TestState_MaybeMoveAbsResourceInstance (0.00s)
    --- PASS: TestState_MaybeMoveAbsResourceInstance/first_move (0.00s)
    --- PASS: TestState_MaybeMoveAbsResourceInstance/noop (0.00s)
=== RUN   TestState_MoveModuleInstance
--- PASS: TestState_MoveModuleInstance (0.00s)
=== RUN   TestState_MaybeMoveModuleInstance
=== RUN   TestState_MaybeMoveModuleInstance/first_move
=== RUN   TestState_MaybeMoveModuleInstance/noop
--- PASS: TestState_MaybeMoveModuleInstance (0.00s)
    --- PASS: TestState_MaybeMoveModuleInstance/first_move (0.00s)
    --- PASS: TestState_MaybeMoveModuleInstance/noop (0.00s)
=== RUN   TestState_MoveModule
=== RUN   TestState_MoveModule/basic
=== RUN   TestState_MoveModule/nested_modules
--- PASS: TestState_MoveModule (0.00s)
    --- PASS: TestState_MoveModule/basic (0.00s)
    --- PASS: TestState_MoveModule/nested_modules (0.00s)
PASS
ok      github.com/hashicorp/terraform/internal/states  0.298s
```

### CODE REVIEW

File: internal/terraform/context_plan.go [hunk 1]
```diff
@@ -826,7 +826,7 @@ func (c *Context) planWalk(config *configs.Config, prevRunState *states.State, o
 			// In refresh-only mode we explicitly don't expect to propose any
 			// actions, but the plan is applyable if the state was changed
 			// in an interesting way by the refresh step.
-			plan.Applyable = !plan.PriorState.ManagedResourcesEqual(plan.PrevRunState)
+			plan.Applyable = !plan.PriorState.ManagedResourcesEqual(plan.PrevRunState) || !plan.PriorState.RootOutputValuesEqual(plan.PrevRunState)
```
Comment (bschaatsbergen) [inline internal/terraform/context_plan.go:829]: As noted in the PR, you could modify this to use `Equal` here. However, I opted to check specifically for root output values, as I felt that switching to `Equal` would be too big of a change.

```suggestion
			plan.Applyable = !plan.PriorState.Equal(plan.PrevRunState)
```

**Comment by liamcervante** [comment]
I think I'm okay with this change. Can you get all the tests passing @bschaatsbergen?

**Comment by jbardin** [comment]
No, data sources are not considered in the comparison, they only really exist in the state at all for historical reasons and don't need to be refreshed.

**Comment by bschaatsbergen** [comment]
> I think I'm okay with this change. Can you get all the tests passing @bschaatsbergen?

Fixed, https://github.com/hashicorp/terraform/pull/35812/commits/73320c586ff37b4a753ff6fdc32c2f5d147a8181

**Review by liamcervante — APPROVED** [review]

**Pull request merged**